### PR TITLE
Support more than 4G of interned strings with StringBackend

### PR DESCRIPTION
**Pull request by shikhar**

### CODE REVIEW

File: src/backend/string.rs [hunk 1]
```diff
@@ -142,12 +141,7 @@ where
     /// If the backend ran out of symbols.
     fn push_string(&mut self, string: &str) -> S {
         self.buffer.push_str(string);
-        let to = self
-            .buffer
-            .as_bytes()
-            .len()
-            .try_into()
-            .expect("ran out of symbols");
```
Comment (shikhar) [inline src/backend/string.rs:150]: this was inaccurate

**Review by Robbepop — COMMENTED** [review]
Thank you for the PR.
This reminds me of one improvement i wanted to make in order to allow for 32-bit counters (as before this PR) but with static safety depending on the chosen symbol type. I think this PR fits as a bridging solution until then.

The problem is that most uesrs don't need more than 2^32 different symbols. So storing 64-bit `usize` (on 64-bit platforms) instead of 32-bit `u32` is a huge waste of memory. Since the `DefaultSymbolType` is a wrapper around `u32` I thought this was fine. However, the default symbol type is not always used so we need a safer API. I already have one in my head. Just needs coding.

**Comment by shikhar** [comment]
@Robbepop this is not the number of symbols, but the length of the concatenated string buffer. That can easily exceed 2^32 even with far less than 2^32 symbols.

**Comment by Robbepop** [comment]
> @Robbepop this is not the number of symbols, but indexes into the concatenated string buffer. That can easily exceed 2^32 even with far less than 2^32 symbols.

This unfortunately is a very good point you made! :thinking:

**Comment by Robbepop** [comment]
The tests are currently failing for this PR since they actively change the memory footprint of the `StringBackend` which is measured in one of the tests. You need to set the expected values there to the ones after the change introduced by this PR to make it pass. I am okay with this since this seems to fix an existing problem.

**Comment by Robbepop** [comment]
> The tests are currently failing for this PR since they actively change the memory footprint of the `StringBackend` which is measured in one of the tests. You need to set the expected values there to the ones after the change introduced by this PR to make it pass. I am okay with this since this seems to fix an existing problem.

If you do this we could merge the PR.

**Comment by shikhar** [comment]
updated!

**Review by Robbepop — APPROVED** [review]
Thanks a lot for this quality improvement of the `StringBackend`!

**Pull request merged**